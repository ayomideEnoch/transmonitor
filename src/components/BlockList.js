import React from 'react'
import PropTypes from 'prop-types'
import {Table} from 'reactstrap'
import Loading from '../loading-bubbles.svg'

let handleClick = (e, props) => {
  // should not prevent default here since we want to jump to the top of
  // transactions list with the html anchor
  if (e.target.id) {
    props.selectBlock(parseInt(e.target.id, 10))
  }
  if (e.target.dataset.block) {
    props.getBlock(parseInt(e.target.dataset.block, 10))
  }
}

export function BlockList (props) {
  if (props.loading) {
    return (
      <div><img src={Loading} alt='Loading...' /></div>
    )
  }

  if (props.blocks.length === 0) return null
  return (
    <Table id='blocks' bordered striped hover responsive size='sm'>
      <caption>List of blocks</caption>
      <thead>
        <tr>
          <th>Block #</th>
          <th>Transactions (with value/total)</th>
          <th>Timestamp</th>
        </tr>
      </thead>
      <tbody onClick={(e) => handleClick(e, props)}>
        {props.blocks.map(block => {
          block.selected = block.number === props.selectedBlock
          return <Block {...block} key={block.number} />
        })}
      </tbody>
    </Table>
  )
}

function Block (block) {
  let selected = block.selected ? 'table-primary' : ''
  let blockNumber = block.totalTransactionsLength !== undefined ? (<span>{block.number}</span>)
    : (<span>
      <a href='#blocks' data-block={block.number}>{block.number}</a>
    </span>)
  let transactions = (block.valueTransactions && block.totalTransactionsLength !== undefined) ? (<span>
    <a href='#transactions' id={block.number}>
      {block.valueTransactions.length}</a> / {block.totalTransactionsLength}
  </span>
  ) : (<span>...</span>)
  return (<tr className={selected}>
    <th>{blockNumber}</th>
    <th>{transactions}</th>
    <th>{block.timestamp ? new Date(block.timestamp * 1000).toLocaleString() : ''}</th>
  </tr>)
}

BlockList.propTypes = {
  blocks: PropTypes.array,
  getBlock: PropTypes.func.isRequired,
  loading: PropTypes.bool.isRequired,
  selectBlock: PropTypes.func.isRequired
}

export default BlockList
