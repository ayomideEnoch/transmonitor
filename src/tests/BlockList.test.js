import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import BlockList from '../components/BlockList';

describe('BlockList', function() {
  let mockPropsLoading = {
    loading: true,
    selectBlock: () => {},
  };

  let mockPropsNoBlock = {
    blocks: [],
    loading: false,
    selectBlock: () => {},
  };

  // the order should be 123458: 3/6, 123457: 1/5, 123456: 2/4
  let mockProps = {
    blocks: [
      {
        number: 123456,
        valueTransactions: [{}, {}],
        totalTransactionsLength: 4,
        timestamp: 1519748998
      },
      {
        number: 123457,
        valueTransactions: [{}],
        totalTransactionsLength: 5,
        timestamp: 1519749008
      },
      {
        number: 123458,
        valueTransactions: [{}, {}, {}],
        totalTransactionsLength: 6,
        timestamp: 1519749108
      },
    ],
    loading: false,
    selectBlock: () => {},
  };

  it('renders without crashing when loading', () => {
    const div = document.createElement('div');
    ReactDOM.render(<BlockList {...mockPropsLoading} />, div);
    ReactDOM.unmountComponentAtNode(div);
  })

  it('renders without crashing after loading', () => {
    const div = document.createElement('div');
    ReactDOM.render(<BlockList {...mockPropsNoBlock} />, div);
    ReactDOM.unmountComponentAtNode(div);
  })

  it('renders null without crashing after loading', () => {
    const div = document.createElement('div');
    ReactDOM.render(<BlockList {...mockPropsNoBlock} />, div);
    ReactDOM.unmountComponentAtNode(div);
  })

  it('render null when no blocks', () => {
    const component = renderer.create(<BlockList {...mockPropsNoBlock} />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  })

  it('render correctly', () => {
    const component = renderer.create(<BlockList {...mockProps} />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  })
});
