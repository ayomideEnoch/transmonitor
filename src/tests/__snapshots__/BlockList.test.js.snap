// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlockList render correctly 1`] = `
<div
  className="table-responsive"
>
  <table
    className="table table-sm table-bordered table-striped table-hover"
    id="blocks"
  >
    <caption>
      List of blocks
    </caption>
    <thead>
      <tr>
        <th>
          Block #
        </th>
        <th>
          Transactions (with value/total)
        </th>
        <th>
          Timestamp
        </th>
      </tr>
    </thead>
    <tbody
      onClick={[Function]}
    >
      <tr
        className=""
      >
        <th>
          <span>
            123458
          </span>
        </th>
        <th>
          <span>
            <a
              href="#transactions"
              id={123458}
            >
              3
            </a>
             / 
            6
          </span>
        </th>
        <th>
          2018-2-28 00:31:48
        </th>
      </tr>
      <tr
        className=""
      >
        <th>
          <span>
            123457
          </span>
        </th>
        <th>
          <span>
            <a
              href="#transactions"
              id={123457}
            >
              1
            </a>
             / 
            5
          </span>
        </th>
        <th>
          2018-2-28 00:30:08
        </th>
      </tr>
      <tr
        className=""
      >
        <th>
          <span>
            123456
          </span>
        </th>
        <th>
          <span>
            <a
              href="#transactions"
              id={123456}
            >
              2
            </a>
             / 
            4
          </span>
        </th>
        <th>
          2018-2-28 00:29:58
        </th>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`BlockList render null when no blocks 1`] = `null`;
